Replace `dim_apply` with `TensorIterator`

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58909 TensorIterator: Reduce serial_for_each static overhead
* **#58656 Replace `dim_apply` with `TensorIterator`**

Ref gh-56794

`dim_apply` is problematic because it calls `Tensor.select` inside of a parallel
region. Instead, replace it with `TensorIterator` by squashing the
apply-dimension. This is similar to the `_dim_apply` function already used by
the sort kernels:

https://github.com/pytorch/pytorch/blob/8c91acc161a742f9f36ab37a55ec46767404b544/aten/src/ATen/native/cpu/SortingKernel.cpp#L27

Differential Revision: [D28776441](https://our.internmc.facebook.com/intern/diff/D28776441)

peterbell10 committed:
Replace `dim_apply` with `TensorIterator`

Ref gh-56794

`dim_apply` is problematic because it calls `Tensor.select` inside of a parallel
region. Instead, replace it with `TensorIterator` by squashing the
apply-dimension. This is similar to the `_dim_apply` function already used by
the sort kernels:

https://github.com/pytorch/pytorch/blob/8c91acc161a742f9f36ab37a55ec46767404b544/aten/src/ATen/native/cpu/SortingKernel.cpp#L27

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace `dim_apply` with `TensorIterator`"


Ref gh-56794

`dim_apply` is problematic because it calls `Tensor.select` inside of a parallel
region. Instead, replace it with `TensorIterator` by squashing the
apply-dimension. This is similar to the `_dim_apply` function already used by
the sort kernels:

https://github.com/pytorch/pytorch/blob/8c91acc161a742f9f36ab37a55ec46767404b544/aten/src/ATen/native/cpu/SortingKernel.cpp#L27

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace `dim_apply` with `TensorIterator`"


Ref gh-56794

`dim_apply` is problematic because it calls `Tensor.select` inside of a parallel
region. Instead, replace it with `TensorIterator` by squashing the
apply-dimension. This is similar to the `_dim_apply` function already used by
the sort kernels:

https://github.com/pytorch/pytorch/blob/8c91acc161a742f9f36ab37a55ec46767404b544/aten/src/ATen/native/cpu/SortingKernel.cpp#L27

[ghstack-poisoned]

mruberry commented:
@ngimel I think you're a better reviewer for this?

peterbell10 committed:
Update on "Replace `dim_apply` with `TensorIterator`"


Ref gh-56794

`dim_apply` is problematic because it calls `Tensor.select` inside of a parallel
region. Instead, replace it with `TensorIterator` by squashing the
apply-dimension. This is similar to the `_dim_apply` function already used by
the sort kernels:

https://github.com/pytorch/pytorch/blob/8c91acc161a742f9f36ab37a55ec46767404b544/aten/src/ATen/native/cpu/SortingKernel.cpp#L27

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace `dim_apply` with `TensorIterator`"


Ref gh-56794

`dim_apply` is problematic because it calls `Tensor.select` inside of a parallel
region. Instead, replace it with `TensorIterator` by squashing the
apply-dimension. This is similar to the `_dim_apply` function already used by
the sort kernels:

https://github.com/pytorch/pytorch/blob/8c91acc161a742f9f36ab37a55ec46767404b544/aten/src/ATen/native/cpu/SortingKernel.cpp#L27

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace `dim_apply` with `TensorIterator`"


Ref gh-56794

`dim_apply` is problematic because it calls `Tensor.select` inside of a parallel
region. Instead, replace it with `TensorIterator` by squashing the
apply-dimension. This is similar to the `_dim_apply` function already used by
the sort kernels:

https://github.com/pytorch/pytorch/blob/8c91acc161a742f9f36ab37a55ec46767404b544/aten/src/ATen/native/cpu/SortingKernel.cpp#L27

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace `dim_apply` with `TensorIterator`"


Ref gh-56794

`dim_apply` is problematic because it calls `Tensor.select` inside of a parallel
region. Instead, replace it with `TensorIterator` by squashing the
apply-dimension. This is similar to the `_dim_apply` function already used by
the sort kernels:

https://github.com/pytorch/pytorch/blob/8c91acc161a742f9f36ab37a55ec46767404b544/aten/src/ATen/native/cpu/SortingKernel.cpp#L27

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -1915,69 +1915,91 @@ void qtopk_kernel(Tensor& values,
+  auto mode_indices_stride = indices.strides()[dim];
+  auto tmp_values_stride = self.strides()[dim];
+
   AT_DISPATCH_QINT_TYPES(self.scalar_type(), "qtopk_cpu", [&] {
```
Other than dispatch macro quantized and non-quantized implementations seem to be exactly the same, can they be refactored to be a template function?

peterbell10 committed:
Update on "Replace `dim_apply` with `TensorIterator`"


Ref gh-56794

`dim_apply` is problematic because it calls `Tensor.select` inside of a parallel
region. Instead, replace it with `TensorIterator` by squashing the
apply-dimension. This is similar to the `_dim_apply` function already used by
the sort kernels:

https://github.com/pytorch/pytorch/blob/8c91acc161a742f9f36ab37a55ec46767404b544/aten/src/ATen/native/cpu/SortingKernel.cpp#L27

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -298,34 +298,60 @@ std::tuple<Tensor&, Tensor&> kthvalue_out_impl_cpu(
+        TensorAccessor<scalar_t, 1> tmp_values(
+            reinterpret_cast<scalar_t*>(data[0] + i * strides[0]),
+            &sizes[dim], &tmp_values_stride);
+        TensorAccessor<scalar_t, 1> tmp_indices(
```
Do you know why this wrong type wasn't caught by the tests? If scalar_t and int64_t have different width, results should have been wrong?

ngimel reviewed (approved):
This looks good, I have a question about why wrong template type for TensorAccessor wasn't caught by the tests. Are we lacking test coverage and some tests should be added?

ngimel commented:
@peterbell10 can you please rebase? I'm getting an error when trying to import this diff
```
RuntimeError: We tried to import new changes for D28735330, but this diff has already been landed.  This could occur if you rebased your stack prior to the landed diff being pushed by ShipIt; in this case, wait until this diff is exported with ShipIt, rebase on master, and try again.
```

peterbell10 committed:
Update on "Replace `dim_apply` with `TensorIterator`"


Ref gh-56794

`dim_apply` is problematic because it calls `Tensor.select` inside of a parallel
region. Instead, replace it with `TensorIterator` by squashing the
apply-dimension. This is similar to the `_dim_apply` function already used by
the sort kernels:

https://github.com/pytorch/pytorch/blob/8c91acc161a742f9f36ab37a55ec46767404b544/aten/src/ATen/native/cpu/SortingKernel.cpp#L27

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28776441).
